### ISSUE TITLE
Incorrect value for 'Start in' field in the generated shortcut.

### ISSUE THREAD

**Issue by noctis0430** [oleg-shilo/wixsharp#878]
The shortcut generated does not have a proper **Start in** value. Expecting the value to be the directory of the target file specified by the **Target** field.

![image](https://user-images.githubusercontent.com/29354176/87113112-191b5900-c2a0-11ea-86e5-6b26be618a99.png)

The following is the code for the shortcut.

```csharp
var shortcut = ExeFileShortcut("Awesome app", @"[INSTALLDIR]\bin\awesome.app.exe", string.empty) { IconFile = @"..\..\resources\images\icon.ico" }
```

Tracing down the source code, it seems to related to `Compiler.GetShortcutWorkingDirectopry()` which is defined as

```csharp
static string GetShortcutWorkingDirectopry(string targetPath)
{
    string workingDir = targetPath;
    var pos = workingDir.LastIndexOfAny(@"\/]".ToCharArray());
    if (pos != -1)
        workingDir = workingDir.Substring(0, pos + 1)
                               .Replace("[", "")
                               .Replace("]", "");
    return workingDir;
}
```

Basically it just removed the square bracket and does not actually resolve `INSTALLDIR`. Is this behavior intended?

**Comment by oleg-shilo** [comment]
Yes it just removes the brackets. This si the convention that MSI is implementing. 
Some WiX entities support attributes that are names of the MSI properties expanded (substituted with real value like path) at runtime. 
But it has to be the name of the property not path including this name.  

In your case the directory property name/id should be (I am guessing) `bin`:

```C#
var shortcut = ExeFileShortcut("Awesome app", @"[INSTALLDIR]\bin\awesome.app.exe", "") 
               {
                   WorkingDirectory = "bin",
                   IconFile = @"..\..\resources\images\icon.ico" 
               }
```
Or

```C#
var shortcut = ExeFileShortcut("Awesome app", @"[bin]\awesome.app.exe", "") 
               {
                   IconFile = @"..\..\resources\images\icon.ico" 
               }
```

**Comment by noctis0430** [comment]
This results in empty value for **Start in**

> ```cs
> var shortcut = ExeFileShortcut("Awesome app", @"[INSTALLDIR]\bin\awesome.app.exe", "") 
>                {
>                    WorkingDirectory = "bin",
>                    IconFile = @"..\..\resources\images\icon.ico" 
>                }
> ```
 
while this one still return **INSTALLDIR\bin\**
 
> ```cs
> var shortcut = ExeFileShortcut("Awesome app", @"[bin]\awesome.app.exe", "") 
>                {
>                    IconFile = @"..\..\resources\images\icon.ico" 
>                }
> ```

Is there anyway to make it expand to full path to the directory?

**Comment by oleg-shilo** [comment]
My assumption was that you do have `bin` (id) directory in you setup. But I could be wrong.

Have a look at https://github.com/oleg-shilo/wixsharp/blob/master/Source/src/WixSharp.Samples/Wix%23%20Samples/Shortcuts/setup.cs sample. 

It sets working dir correctly. I tested it before answering your question. Also have a look at the generated wxs. It may give you some ideas.

**Comment by noctis0430** [comment]
Sorry for getting back to you late. I've been busy at my workplace for the last few days.

> Have a look at https://github.com/oleg-shilo/wixsharp/blob/master/Source/src/WixSharp.Samples/Wix%23%20Samples/Shortcuts/setup.cs sample.

Your updated samples do help me to identify the issue in my codes. 

The issue is that when working directory is not set on the  `ExeFileShorcut` instance, `Compiler.GetShortcutWorkingDirectory()` will use `ExeFileShortcut.Target` property to try resolve the working directory. In my case, that will be `[INSTALLDIR]\bin\awesome.app.exe` being resolved to `INSTALLDIR\bin\`. However, the directory ID generated in the wxs is `INSTALLDIR.bin` instead of `INSTALLDIR\bin\`. I guess that's why it cannot be resolved correctly when the shortcut is created.

Once I manually set the working directory like so, the issue goes away.

```cs
var shortcut = ExeFileShortcut("Awesome app", @"[INSTALLDIR]\bin\awesome.app.exe", string.empty) { WorkingDirectory="INSTALLDIR.bin", IconFile = @"..\..\resources\images\icon.ico" };
```

Could this be another bug?

**Comment by oleg-shilo** [comment]
Not sure it can be considered a bug. Though it can be challenged.  
If you did not specify WorkingDir then it does not matter what it is at runtime. WixSharp attempts so solve it by removing the file name from the path but it would only work if the user specify it in a form of `<dir_id>\<file_name>` For example:

```C#
ExeFileShortcut("Awesome app", @"[INSTALLDIR.bin]\awesome.app.exe",...
```

If the user opted to use indirect directory id (e.g. `<dir_id>\<subdir_name>\<file_name>`) then this algorithm will not work and it is netter to set WorkingDir explicitly as you did.

I have reflected it in the API documentation. 

![image](https://user-images.githubusercontent.com/16729806/87551045-20ed5a00-c6f3-11ea-8a20-9413fcac4d9c.png)

**Comment by oleg-shilo** [comment]
And of course I misspelled "more" as "mmore" :)

**Comment by noctis0430** [comment]
> And of course I misspelled "more" as "mmore" :)

Not an issue there. The message has been conveyed clearly 😄
